keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,474 @@
+
+The approach of keeping the information in the status of
+the queues will allow to expose the information without an extra cost of new
+API requests, as the statuses are already updated on changes for the number of
+pending workloads.
+
+In order to keep the size of the information constrained only the head of the
+queue of pending workloads will be exposed.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user of Kueue with LocalQueue visibility I would like to know the
+position of my workload in the local queue. Knowing the position would allow me
+to compute the estimate arrival time (ETA) of my workload.
+
+I would like to be able to get this information by inspecting the local queue
+status.
+
+#### Story 2
+
+As a user of Kueue with ClusterQueue visibility I would like to know
+the position of my workload in the cluster queue. This will allow me to answer
+users questions about their workloads. Merging the information exposed by
+individual local queues is not enough, because they may be showing inconsistent
+data due to batching of their updates.
+
+I would like to be able to get this information by inspecting the cluster queue
+status.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Time and memory consuming computation of top pending workloads
+
+Currently, we organize the cluster queue as a heap, thus only the top of the
+heap is readily available. If we want to get to know the top N pending workloads
+we may need to copy the heap, sort and select N, which is time and memory
+consuming.
+
+In order to mitigate this risk we may need to migrate from using a heap to
+red-black trees.
+
+#### Too large local queue object
+
+As the number of pending workloads is arbitrarily large there is a risk that the
+status information about the workloads may exceed the etcd limit of 1.5Mi on
+object size. In order to allow feature extensions of the structure we
+should assume not more that 500Ki is used.
+
+In order to mitigate this risk we put a constraint on the number of exposed
+pending workloads. We limit the number to 1000.
```

Comment:
For objects that we expect to number in thousands or more in the cluster - such as LocalQueue may if there's one per user in a large organization, or one per experiment - I don't think it's a good idea to aim for nearly maxing out the object size. In such setup there'll likely be a need to list and watch queues, and watching large objects is much more expensive.

I think it would be good to keep the default size of objects much lower than that, even with high churn. If we can parametrize the number of pending workloads and set the default to some low number, that would allow the feature to both be enabled by default, and scalable by default. Users can then set a higher value if they prefer.

Reply:
Yeah, I'm hesitant how to come up with estimations for the default. The etcd limit is something tangible we can use for estimations relatively easily. The situations with many LocalQueues, many watches on them and high churn, has many variables which are weirdly connected, that does not give me a good basis for estimations. 

Also, I think of two scenarios:

Reply:
I think there is a natural analogy with Jobs, where for `Indexed Jobs` and `Backoff Limit Per Index` features the size of the status can be large, so we have limits for fields that control it. The approach we took was to setup limits so that it is roughly 500k object size as max. 

I think the considerations between Jobs and LocalQueues are very similar: 
- The object size may grow with new functionalities, for example, we may want to add the ETA field per each workload,
- The worst case scenario is unlikely in practice, since it assumes the lengths namespaces and names to be of max length of 253 characters.
- Many instances of LocalQueue (or Job) are possible, they both can be watched

For reference calculations: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#the-job-object-too-big

Reply:
Following this strategy, I’m now leaning to set the default at 1000, which translates to roughly 500k.

Reply:
FWIW, the number of LocalQueues is constrained by the number of namespaces, which is up to 10k: https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds

Reply:
Another aspect here, that makes the LocalQueues different from Jobs is that here we need to take the read lock for the duration of the iteration over the global cluster queue. Thus, setting the lower limit for local queues makes sense actually.

Reply:
> Another aspect here, that makes the LocalQueues different from Jobs is that here we need to take the read lock for the duration of the iteration over the global cluster queue. Thus, setting the lower limit for local queues makes sense actually.

I have updated the proposal to take a periodic in-memory snapshot of ClusterQueue, this is to avoid taking a prolonged read lock on the ClusterQueue while generating the structure for LocalQueue.

Reply:
I've added a section in Design Details on setting the defaults, and updated the sections on risks. @aleksandra-malinowska PTAL. In particular, I'm wondering if there is a way to translate the risk related to watches into some estimations. For now, I propose to set the limit to 1000 (following the Job Backoff Limit Per Index), but the default to 100, to reduce the chance of issues as the feature is enabled by defaults.